File: sdk/python/src/trident_indexer/webhook.py
"""Webhook signature verification helpers for Trident (issue #452).

Receivers call :func:`verify_signature` on every incoming webhook request to
confirm the delivery is authentic and within the replay-protection window.

Signing scheme (for offline validation)::

    import hmac, hashlib
    message  = f"{timestamp}.{body}".encode()
    mac      = hmac.new(secret.encode(), message, hashlib.sha256)
    expected = "sha256=" + mac.hexdigest()

The ``X-Trident-Signature`` header may contain two space-separated signatures
during a secret rotation overlap window.  Pass your current active secret —
the function checks all tokens and returns ``True`` if any one matches.
"""

from __future__ import annotations

import hashlib
import hmac
import time
from typing import Union

__all__ = [
    "DEFAULT_TOLERANCE_SECONDS",
    "WebhookVerificationError",
    "verify_signature",
    "compute_signature",
]

#: Recommended replay-protection window in seconds (5 minutes).
DEFAULT_TOLERANCE_SECONDS: int = 300


class WebhookVerificationError(Exception):
    """Raised by :func:`verify_signature` when verification fails.

    The ``reason`` attribute carries a human-readable explanation for logging.
    Do **not** surface it in HTTP responses to callers.
    """

    def __init__(self, reason: str) -> None:
        super().__init__(f"webhook signature verification failed: {reason}")
        self.reason = reason


def compute_signature(timestamp: int, body: Union[str, bytes], secret: str) -> str:
    """Return the ``sha256=<hex>`` signature for *body* at *timestamp*.

    This is the reference implementation — use it to produce test vectors or
    to sign outgoing requests in your own test server.
    """
    if isinstance(body, str):
        body = body.encode()
    message = f"{timestamp}.".encode() + body
    mac = hmac.new(secret.encode(), message, hashlib.sha256)
    return "sha256=" + mac.hexdigest()


def verify_signature(
    body: Union[str, bytes],
    signature: str,
    timestamp: str,
    secret: str,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
) -> None:
    """Verify an incoming Trident webhook delivery.

    :param body: Raw request body — read it **before** parsing JSON.
    :param signature: Value of the ``X-Trident-Signature`` header.
    :param timestamp: Value of the ``X-Trident-Timestamp`` header (Unix seconds).
    :param secret: Your webhook subscription secret (starts with ``whsec_``).
    :param tolerance_seconds: Maximum delivery age in seconds.  Pass ``0`` to
        disable the timestamp check (not recommended in production).
        Defaults to :data:`DEFAULT_TOLERANCE_SECONDS` (300 s).

    :raises WebhookVerificationError: on any verification failure.
    :raises WebhookVerificationError: if the timestamp is outside the tolerance window.
    """
    # 1. Parse and range-check the timestamp.
    try:
        ts = int(timestamp)
    except (ValueError, TypeError):
        raise WebhookVerificationError(
            "X-Trident-Timestamp is missing or not a valid Unix second"
        )
    if ts <= 0:
        raise WebhookVerificationError(
            "X-Trident-Timestamp must be a positive Unix second"
        )
    if tolerance_seconds > 0:
        age = abs(int(time.time()) - ts)
        if age > tolerance_seconds:
            raise WebhookVerificationError(
                f"timestamp is {age}s old (tolerance {tolerance_seconds}s) — possible replay"
            )

    # 2. Compute the expected signature.
    expected = compute_signature(ts, body, secret)

    # 3. Accept any token in a space-separated signature header.
    for token in signature.split():
        if hmac.compare_digest(token, expected):
            return

    raise WebhookVerificationError("signature does not match")

File: sdk/go/webhook.go
package trident

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// DefaultWebhookToleranceSeconds is the recommended replay-protection window.
// Reject any webhook whose X-Trident-Timestamp differs from the current wall
// clock by more than this value. Five minutes matches Stripe's convention and
// is wide enough to absorb normal network and clock-skew jitter while keeping
// the replay window short.
const DefaultWebhookToleranceSeconds int64 = 300

// WebhookVerificationError is returned by VerifyWebhookSignature when
// verification fails. The Reason field carries a human-readable explanation
// suitable for logging (never send it to the end user).
type WebhookVerificationError struct {
	Reason string
}

func (e *WebhookVerificationError) Error() string {
	return "webhook signature verification failed: " + e.Reason
}

// VerifyWebhookSignature verifies that an incoming webhook delivery is
// authentic and within the replay-protection window.
//
// Parameters:
//   - body:      the raw request body bytes (do not parse before verifying)
//   - signature: the value of the X-Trident-Signature header
//   - timestamp: the value of the X-Trident-Timestamp header (Unix seconds, as a string)
//   - secret:    your webhook subscription secret (starts with "whsec_")
//   - toleranceSecs: maximum age of the delivery in seconds; pass
//     DefaultWebhookToleranceSeconds (300) for the recommended window, or 0
//     to skip the timestamp check (not recommended in production)
//
// Returns nil on success, *WebhookVerificationError on any failure.
//
// Signing scheme (for offline validation):
//
//	message  = fmt.Sprintf("%d.%s", timestampUnix, rawBody)
//	mac      = HMAC-SHA256(key=secret, message=message)
//	expected = "sha256=" + hex.EncodeToString(mac)
//
// The X-Trident-Signature header may contain two space-separated signatures
// during a secret rotation overlap window; verification passes if either one
// matches. Always verify against your current active secret — the server sends
// both during rotation so you have time to swap your key.
func VerifyWebhookSignature(body []byte, signature, timestamp, secret string, toleranceSecs int64) error {
	// 1. Parse and range-check the timestamp.
	// strconv.ParseInt, not fmt.Sscanf: Sscanf stops at the first non-digit and
	// reports success, so "1700000000junk" would parse as 1700000000. The other
	// SDKs reject trailing garbage, and this one must agree with them.
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil || ts <= 0 {
		return &WebhookVerificationError{Reason: "X-Trident-Timestamp is missing or not a valid Unix second"}
	}
	if toleranceSecs > 0 {
		age := time.Now().Unix() - ts
		if age < 0 {
			age = -age // tolerate small clock skew in both directions
		}
		if age > toleranceSecs {
			return &WebhookVerificationError{
				Reason: fmt.Sprintf("timestamp is %d seconds old (tolerance %d s) — possible replay", age, toleranceSecs),
			}
		}
	}

	// 2. Compute the expected signature.
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = fmt.Fprintf(mac, "%d.%s", ts, body)
	expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	// 3. Accept any token in a space-separated signature header.
	for _, token := range strings.Fields(signature) {
		if subtle.ConstantTimeCompare([]byte(token), []byte(expected)) == 1 {
			return nil
		}
	}
	return &WebhookVerificationError{Reason: "signature does not match"}
}
